Better SecretStore mocking

zanderz opened the pull request:
Ran libkb and engine tests on Linux and Windows.
Learned some things about sharing test code across packages in Go.
Thought of moving the mock activate/deactivate calls to the global test setup/teardown functions - what do you guys think?
@maxtaco @patrickxb

maxtaco commented:
Thanks Steve, I took a look. I think a much cleaner way to do this would be to dangle the secret store off of the `G` object, which is now only local to tests.   On OSX populate it with a real secret store.  On non-OSX, in tests, populate it with a mock local secret store.  

I worry the strategy used here of inherently manipulating global state (and unmanipulating it when the test is over) is race-prone and  will cause problems down the road.  We've seen this problem historically and have solved it by keeping all test-state (like mock API replies, for instance) dangling off of G.

maxtaco commented:
@patrickxb what do you think about this?

I was kind of surprised that the secret store isn't dangling off of G when I was looking around a few days ago to see how `LoginWithSecretStore` was working.

patrickxb commented:
Yes, I think putting the secret store in G and having it populated with a mock automatically during testing would make testing easier as we could always assume there was a secret store available.  Would need to make sure it is deleted on Logout, user change.

I believe @akalin-keybase's original secret store integration was designed so that you could create a new one with a username whenever you wanted a secret store.

zanderz commented:
Should I try to revive @akalin-keybase 's efforts here or just put what we have into G? I'll assume the latter unless I hear otherwise.

maxtaco commented:
I vote the latter!

zanderz commented:
Hey guys, thanks for the comments. I thought I was all clear but now that I've come back to this today it seems less so. I take it when you refer to this proposed secret store member of the global context, this would be some new interface, instead of the current thing called SecretStore, which are just temporary accessors that manipulate global state? This new thing would have its methods invoked instead of creating NewSecretStore() all the time for setting and retrieving values? Or do you intend something else? Thanks.

maxtaco commented:
Looks like you need to put both a `SecretStore` interface.  You might make the follow changes:

* Dangle a `SecretStoreContext` off of `SecretStore`; meaning the `SecretStore` interface gets a `GetSecretStoreContext` method
* Add an `accountName` parameters to `StoreSecret`, `ClearSecret` and `RetreiveSecret`

@akalin-keybase do these changes seem reasonable?  The high level goal is to allow for a mock secret store to dangle off of `GlobalContext` in the case of cross-platform testing.

akalin-keybase commented:
This sounds good to me. I don't think I had a strong reason for baking the username into the `SecretStore` interface. However, with the new proposal you'd have to plumb through the username to whatever currently uses the secret store, which may or may not be a lot of work. If it is a lot of work, it may be easier to just put a new interface, say, `SecretStoreFactory` into `GlobalContext` that vends `SecretStore` objects. Either way!

zanderz commented:
Hey guys, I thought I would seek another review before proceeding, since by now merging back to master may not be trivial. Also, for this I will seek help doing IOS and Android builds before merging to master.

maxtaco commented:
Took a quick look on my phone and two immediate things pop out: C/I broken
on Mac with compiler errors; also in this pr and everywhere else, please
never introduce new references to 'G'; plumb through instead via dependency
injection.

We also need to test heavily by hand on OS X to make sure this doesn't
break anything.

Also why did you get rid of the interface hierarchy with sender and
receivers?

On Monday, March 7, 2016, Steve Sanders <notifications@github.com> wrote:

> Hey guys, I thought I would seek another review before proceeding, since
> by now merging back to master may not be trivial. Also, for this I will
> seek help doing IOS and Android builds before merging to master.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/keybase/client/pull/2179#issuecomment-193145878>.
>

zanderz commented:
Sorry about that. G is ok in tests though right? So I took GetTerminalPrompt() back out of the interface; didn't seem worth marshaling it across RPC, unless you guys think otherwise.

The senders and receivers weren't being used for anything anymore. This refactoring graduates the thing named SecretStore from a per-user key to a multi-user interface, now part of the global context. Instead of building a sender/receiver and discarding it after, you just store and query straight to SecretStore with the username. I thought that's what you guys were talking about, but of course it doesn't have to be that way.

I won't merge with a CI failure but it is a lot easier to let it build on Mac.

maxtaco commented:
Actually it's especially important in tests not to have a global `G` since in tests in when we find ourselves having multiple `G`s that are actually different.   Can you show me an example of a `G` that you need global?

maxtaco commented:
OK on the sender/receiver stuff BTw!

patrickxb commented on go/libkb/passphrase_helper.go:
```diff
@@ -124,7 +124,7 @@ func DefaultPassphraseArg() keybase1.GUIEntryArg {
 			},
 			StoreSecret: keybase1.Feature{
-				Allow:        HasSecretStore(),
+				Allow:        G.SecretStore != nil,
```
I think you'll have to pass GlobalContext into this function to avoid using capital-G G.

patrickxb commented on go/libkb/secret_store_none.go:
```diff
@@ -5,19 +5,63 @@
+
+	ret, ok := tss.secretStoreNoneMap[accountName]
+
+	G.Log.Debug("| TestSecretStore::RetrieveSecret(%d)", len(ret))
```
can this use the TestContext's Global Context?

patrickxb commented on go/libkb/globals.go:
```diff
@@ -441,7 +443,7 @@ type Contexitifier interface {
 
 func (g *GlobalContext) GetConfiguredAccounts() ([]keybase1.ConfiguredAccount, error) {
-	return GetConfiguredAccounts(g)
+	return GetConfiguredAccounts(g, g.SecretStore)
```
Do we need this?  Can `GetConfiguredAccounts()` be on `SecretStore`?  So then it would be `e.G().SecretStore.GetConfiguredAccounts()`.

patrickxb replied:
Or is e.G().SecretStore for just one user?

zanderz commented:
Thanks @patrickxb , missed a couple of Gs.
@maxtaco , I realize maybe what you mean is that if we kept some kind of a sender/receiver, less of this code would be touched? Maybe that makes more sense?

maxtaco commented:
@zanderz yes that I what I meant....

akalin-keybase commented on go/libkb/skb.go:
```diff
@@ -770,14 +766,14 @@ func (s *SKB) unlockPrompt(arg SecretKeyPromptArg, which string, secretStore Sec
 
 // func (s *SKB) PromptAndUnlock(lctx LoginContext, reason, which string, secretStore SecretStore, ui SecretUI, lksPreload *LKSec, me *User) (ret GenericKey, err error) {
-func (s *SKB) PromptAndUnlock(arg SecretKeyPromptArg, which string, secretStore SecretStore, lksPreload *LKSec, me *User) (ret GenericKey, err error) {
+func (s *SKB) PromptAndUnlock(arg SecretKeyPromptArg, which string, username NormalizedUsername, lksPreload *LKSec, me *User) (ret GenericKey, err error) {
```
since you're already passing in `me`, you don't need to pass in `username` separately

akalin-keybase commented on go/engine/deprovision_test.go:
```diff
@@ -224,7 +214,7 @@ func TestCurrentDeviceRevoked(t *testing.T) {
 	fu := NewFakeUserOrBust(tc.T, "dpr")
 	arg := MakeTestSignupEngineRunArg(fu)
-	arg.StoreSecret = libkb.HasSecretStore()
+	arg.StoreSecret = tc.G.SecretStore != nil // generally true
```
Unfortunately, we lose test coverage of the path where `SecretStore` is nil, which I think we still need to support. Perhaps have a separate test for that case, and factor out the common code?

akalin-keybase commented:
took a quick look, had a couple of comments.

zanderz commented:
Thanks guys, I changed a bunch of stuff based on all your comments. Several deprovision tests run both with and without the secret store, some global context usages were cleaned up, and I backed off on changing the SecretStore interface. Instead, I added SecretStoreAll, leaving SecretStore to be per-user. Causes less churn. If this passes CI I will rebase to master and see about trying it on Mac.

patrickxb commented on go/libkb/secret_store.go:
```diff
@@ -64,9 +92,13 @@ func GetConfiguredAccounts(c SecretStoreContext) ([]keybase1.ConfiguredAccount,
-	return secretStore.ClearSecret()
+	return g.SecretStoreAll.ClearSecret(username)
+}
+func NewTestSecretStoreAll(c SecretStoreContext, g *GlobalContext) SecretStoreAll {
```
Can this function go in secret_store_test.go so it is only compiled into test builds?

zanderz replied:
I have been going around in circles on this. It must be accessible from test_common.go, I believe mainly so tests in other packages can use it, so it can't be compiled test-only. I will try solving the osx breakage by putting a no-op NewTestSecretStoreAll() in the darwin file, same on Android.

patrickxb commented:
looking good...CI seems to be failing, however.  I added a new comment about NewTestSecretStoreAll.

akalin-keybase commented on go/engine/deprovision_test.go:
```diff
@@ -41,15 +41,13 @@ func getNumKeys(tc libkb.TestContext, fu FakeUser) int {
-	tc := SetupEngineTest(t, "deprovision")
-	defer tc.Cleanup()
-
+func assertDeprovisionWithSetup(t *testing.T, tc libkb.TestContext) {
```
can pass in just `tc`, and use `tc.T` instead of `t`

akalin-keybase commented on go/engine/deprovision_test.go:
```diff
@@ -126,15 +124,27 @@ func TestDeprovision(t *testing.T) {
 	arg := MakeTestSignupEngineRunArg(fu)
-	arg.StoreSecret = libkb.HasSecretStore()
+
+	arg.StoreSecret = tc.G.SecretStoreAll != nil // generally true in tests
```
this comment doesn't really make sense anymore since you're explicitly calling this function with and without SecretStoreAll...suggest just removing

akalin-keybase commented on go/engine/deprovision_test.go:
```diff
@@ -126,15 +124,27 @@ func TestDeprovision(t *testing.T) {
+	assertDeprovisionWithSetup(t, tc)
+}
+
+func assertDeprovisionLoggedOut(t *testing.T, tc libkb.TestContext) {
```
just `tc` here, too

akalin-keybase commented on go/engine/deprovision_test.go:
```diff
@@ -126,15 +124,27 @@ func TestDeprovision(t *testing.T) {
 	tc := SetupEngineTest(t, "deprovision")
 	defer tc.Cleanup()
+	if tc.G.SecretStoreAll == nil {
+		t.Fatalf("Need a secret store for this test")
```
`t.Fatal` (here and others)

akalin-keybase commented:
lgtm after nits and CI and patrick's comment
